src/Avalonia.Controls.DataGrid/Utils/ReflectionHelper.cs

```diff
@@ -352,19 +352,37 @@ internal static PropertyInfo GetPropertyOrIndexer(this Type type, string propert
                 return null;
             }
 
-            PropertyInfo indexer = null;
             string stringIndex = propertyPath.Substring(1, propertyPath.Length - 2);
-            indexer = FindIndexerInMembers(type.GetDefaultMembers(), stringIndex, out index);
+            var indexer = FindIndexerInMembers(type.GetDefaultMembers(), stringIndex, out index);
             if (indexer != null)
             {
                 // We found the indexer, so return it.
                 return indexer;
             }
 
-            if (typeof(IList).IsAssignableFrom(type))
+            var elementType = type.GetElementType();
+            if (elementType == null)
+            {
+                var genericArguments = type.GetGenericArguments();
+                if (genericArguments.Length == 1)
+                {
+                    elementType = genericArguments[0];
+                }
+            }
+
+            if (elementType != null)
             {
                 // If the object is of type IList, try to use its default indexer.
-                indexer = FindIndexerInMembers(typeof(IList).GetDefaultMembers(), stringIndex, out index);
```

Comment:
Previously "IList" would return "object" indexer return type, that isn't usable in our case.